src/extensions/resolveSource.ts

```diff
@@ -10,15 +14,24 @@ export interface RegistryEntry {
   icons?: { [key: string]: string };
   labels: { [key: string]: string };
   versions: { [key: string]: string };
+  updateWarnings?: { [key: string]: UpdateWarning[] };
+}
+
+export interface UpdateWarning {
+  from: string;
+  description: string;
+  action?: string;
 }
 
 /**
  * Gets the source for a given extension name and version from the official extensions registry
  *
  * @param extensionName the name, or name@version of the extension to get the ExtensionSource for
+ * @param startVersion The version to display updateWarnings for.
+ *                      If not present, updateWarnings won't be shown.
  * @returns the source corresponding to extensionName in the registry
  */
-export async function resolveSource(extensionName: string): Promise<string> {
+export async function resolveSource(extensionName: string, startVersion?: string): Promise<string> {
```

Comment:
This feels like an odd pattern to me. Why would an internal function called "resolveSource" have an optional parameter that may require user intervention? If I (accidentally) used this in a script or elsewhere and provided a "currentVersion", I would be very confused by this behavior (and, it's impossible to override any warning via options in this current state, say if `--force`  were to be allowed to skip this confirmation).

I think a different refactor is necessary to have this make sense - and I believe this function name is misleading (just a tad). This doesn't return a source "object", it returns a source "URL". We can do better.

A light-touch way of having this make a little more sense would be to have this function have a better return value, maybe `RegistryEntry` (or something similar - it may be necessary to have something different, maybe `TargetRegistryEntry` that has `sourceName`, `requestedVersion`, `foundVersion`, `updateWarnings` or something... exercise for the reader...))

Then, in `ext-update.ts`, that `RegistryEntry` (or something) can be passed to `checkForUpdateWarnings` that does the additional check.

All in all, burying the user confirmation this deep will make it tricky in the future, so let's see if a better pattern can be discovered.

Reply:
Refactored this to avoid side effects - lmk if this looks better now.